strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Bosonic circuit operations"""
+# pylint: disable=duplicate-code,attribute-defined-outside-init
+import itertools as it
+import numpy as np
+from scipy.linalg import block_diag
+import thewalrus.symplectic as symp
+
+from strawberryfields.backends.bosonicbackend import ops
+
+
+# Shape of the weights, means, and covs arrays.
+def w_shape(num_modes, num_gauss):
+    r"""Calculate total number of weights. Assumes same number of weights per mode.
+
+    Args:
+        num_modes (int): number of modes
+        num_gauss (int): number of gaussian peaks per mode
+
+    Returns:
+        int: number of weights for all Gaussian peaks in phase space
+    """
+    return num_gauss ** num_modes
+
+
+def m_shape(num_modes, num_gauss):
+    r"""Shape of array of mean vectors. Assumes same number of weights per mode.
+
+    Args:
+        num_modes (int): number of modes
+        num_gauss (int): number of gaussian peaks per mode
+
+    Returns:
+        tuple: (number of weights, number of quadratures)
+    """
+    return (num_gauss ** num_modes, 2 * num_modes)
+
+
+def c_shape(num_modes, num_gauss):
+    r"""Shape of array of covariance matricess. Assumes same number of weights per mode.
+
+    Args:
+        num_modes (int): number of modes
+        num_gauss (int): number of gaussian peaks per mode
+
+    Returns:
+        tuple: (number of weights, number of quadratures, number of quadratures)
+    """
+    return (num_gauss ** num_modes, 2 * num_modes, 2 * num_modes)
+
+
+def to_xp(num_modes):
+    r"""Provides array of indices to order quadratures as all x
+    followed by all p, starting from (x1,p1,...,xn,pn) ordering.
+
+    Args:
+        num_modes (int): number of modes
+
+    Returns:
+        array: quadrature ordering for all x followed by all p
+    """
+    return np.concatenate((np.arange(0, 2 * num_modes, 2), np.arange(0, 2 * num_modes, 2) + 1))
+
+
+def from_xp(num_modes):
+    r"""Provides array of indices to order quadratures as (x1,p1,...,xn,pn)
+    starting from all x followed by all p.
+
+    Args:
+        num_modes (int): number of modes
+
+    Returns:
+        list: quadrature ordering for (x1,p1,...,xn,pn)
+    """
+    perm_inds_list = [(i, i + num_modes) for i in range(num_modes)]
+    perm_inds = [a for tup in perm_inds_list for a in tup]
+    return perm_inds
+
+
+def update_means(means, X, perm_out):
```

Comment:
Suggest adding a specific test case for these helper functions

Reply:
The apply_channel method directly applies these functions and we test that.